Add lesson actions block settings

renatho opened the pull request:
Part of #3821

### Changes proposed in this Pull Request

* It introduces the settings to the Lesson Actions, including the Reset Lesson action.
  * It persists the button state while it's being toggled until the user refreshes the page.

### Testing instructions

* Go to the lesson editor.
* Add the _Lesson actions_ block.
* Edit the _Reset lesson_ button.
* In the _Lesson actions_ sidebar settings, disable the _Reset lesson_.
* Enable the button again, and make sure the previous settings continue in the button.

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

<img width="304" alt="Screen Shot 2021-01-12 at 18 15 20" src="https://user-images.githubusercontent.com/876340/104375450-25f46000-5502-11eb-8d3b-18dfa4e4dbec.png">

donnapep commented:
Please note that the text for the block descriptions can be found at the bottom of the [issue description](https://github.com/Automattic/sensei/issues/3821). I updated [this comment](https://github.com/Automattic/sensei/issues/3821#issuecomment-754115899) to indicate that as well.

renatho commented:
> Please note that the text for the block descriptions can be found at the bottom of the [issue description](https://github.com/Automattic/sensei/issues/3821). I updated [this comment](https://github.com/Automattic/sensei/issues/3821#issuecomment-754115899) to indicate that as well.

Thanks, @donnapep! Updated in the other PR: https://github.com/Automattic/sensei/pull/3886/commits/af769dd08ff76af35aeb12c8627b6c2b21f387b9 ;)

gkaragia commented on assets/blocks/lesson-actions/lesson-actions-block/block.json:
```diff
@@ -3,5 +3,13 @@
+  },
+  "attributes": {
+    "activeBlocks": {
+      "type": "object",
```
Maybe an array is more intuitive here?

renatho replied:
Yep! I agree with you that an array would be more intuitive here. I started implementing as array, but considering we toggle only some of the blocks, using an object end up making more sense, and the code is cleaner.

But getting your point, I renamed it to `toggledBlocks`. WDYT? Does it make more sense now?

gkaragia replied:
Not sure to be honest. I still think that it's more intuitive to add/remove a block to an array of blocks for such functionality and an array seems more appropriate for a collection of something. But it might be just me 😄

renatho replied:
I get your point. But I believe if we use the blocks as an array here, the code would become clunky. We would need to add all the possible blocks to the attribute (even we don't wanna toggle most of them), and trust in this array as source of truth to list the blocks. See this [part](https://github.com/Automattic/sensei/pull/3887/files#diff-fc8a7b975ce7ffe6080e2288eee909c3062feae71a5bd6f3a1a8899f9cb87932R118-R120), for example.

gkaragia commented on assets/blocks/lesson-actions/lesson-actions-block/settings.js:
```diff
@@ -0,0 +1,24 @@
+					false !== activeBlocks[ 'sensei-lms/button-reset-lesson' ]
+				}
+				onChange={ toggleBlock( 'sensei-lms/button-reset-lesson' ) }
+				label={ __( 'Reset lesson', 'sensei-lms' ) }
```
I find the label a bit confusing. We could rename this to "Enable 'Reset lesson'" or "'Reset lesson' enabled"

renatho replied:
I don't know if we need that, considering we have the "Additional Actions" title. WDYT @donnapep and @yscik?

donnapep replied:
Taking a look at other core blocks like _Gallery_ or _Cover_, they don't prepend "Enable" to the labels. Implicitly most people understand that a toggle enables or disables something.

gkaragia commented on assets/blocks/lesson-actions/lesson-actions-block/edit.js:
```diff
@@ -1,24 +1,100 @@
+	 *
+	 * @return {Function} Function to toggle the block.
+	 */
+	const toggleBlock = ( blockName ) => ( on ) => {
```
Do you think it would be better to put this on its own hook? 

We could also make the settings dumper if we created the `toggleBlock` callback here and updated them to have arguments `isResetActive` and `onResetChange` or similar.

renatho replied:
Good idea! It separated this concern, making the `EditLessonActionsBlock` cleaner, and making the setting dumber. =)
Thank you for the suggestion!

gkaragia reviewed (commented):
Looks good works well! Left a few comments.

renatho commented on assets/blocks/lesson-actions/lesson-actions-block/edit.js:
```diff
@@ -1,24 +1,100 @@
-	[ 'sensei-lms/button-reset-lesson', {} ],
+import { LessonActionsBlockSettings } from './settings';
+
+const allowedBlocks = [
```
Yep! Makes sense! https://github.com/Automattic/sensei/pull/3887/commits/8b3e195513391c39e0ee8b1f317e5d6638f99d0a

gkaragia reviewed (approved):
Looks good! We can always update the toggle label later if we find a more fitting name.
